### PR TITLE
Funnel-Transformer

### PR DESCRIPTION
**Pull request by mvxxx**
Implementation of https://arxiv.org/abs/2006.03236

### CODE REVIEW

**Comment by mvxxx** [comment]
@googlebot I signed it!

**Comment by mvxxx** [comment]
@googlebot I consent.

**Comment by syzymon** [comment]
@googlebot I consent.

**Comment by shadowatyyy** [comment]
@googlebot I consent.

**Comment by syzymon** [comment]
@googlebot I fixed it.

File: trax/models/research/funnel_transformer.py [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+  return tl.Serial(
+      tl.Branch(
+          tl.Fn('select_cls_token', lambda x: x[:, :1, :]),
+          tl.Serial(
+              tl.Fn('rest_tokens', lambda x: x[:, 1:, :]),
+              pool_layer(pool_size, strides)
+          )
+      ),
+      tl.Concatenate(axis=1)
+  ) if separate_cls else pool_layer(pool_size, strides)
```
Comment (sebastianjaszczur) [inline trax/models/research/funnel_transformer.py:47]: I would prefer to see here if-else statement rather than conditional expression; this Serial is quite long anyway. Can you split it?

File: trax/models/research/funnel_transformer.py [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+
+
+def _Upsample(short, masks, long):
+  factor = -(-long.shape[1] // short.shape[1])  # ceil division
```
Comment (sebastianjaszczur) [inline trax/models/research/funnel_transformer.py:51]: I don't think this works as upsampling; short counterexample below. I think it would be reasonable just to assert that input length is always divisible by pool_size wherever we use pooling (input length is usually a power of two anyway). Usually the input length in training/evaluation is a power of 2 anyway, I think.

Counterexample: input_size=31, pool_size=2, single decoding block. AvgPool with default padding ("valid", not "same") will produce tensor of length 15. After this downsampling block we will have upsampling with factor = 3. This is out of sync with pool_size - we will get tokens mapped to wrong positions in "short.repeat".
While this counterexample would be fixed by changing the padding to "same", I have a feeling that we should just assert that input length at every stage will be divisible by pool_size. What do you think?

File: trax/models/research/funnel_transformer.py [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+  new_vecs = long + short.repeat(factor, axis=1)[:, :long.shape[1], :]
+  new_masks = masks.repeat(factor, axis=-1)[:, :, :, :long.shape[1]]
```
Comment (sebastianjaszczur) [inline trax/models/research/funnel_transformer.py:53]: I think masks shouldn't be upsampled like this - we could just use original masks instead of downsampling and upsampling them, which may introduce errors.

File: trax/models/research/funnel_transformer.py [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+def _Upsampler():
+  return tl.Fn('Upsampler', _Upsample, n_out=2)
+
+
+def _FunnelBlock(d_model, d_ff, n_heads,
+                 dropout, dropout_shared_axes, mode, ff_activation,
+                 pool_layer, pool_size, strides, separate_cls):
+  """Internal funnel block. On input it takes (activations, masks).
+
+  Args:
```
Comment (sebastianjaszczur) [inline trax/models/research/funnel_transformer.py:66]: Can you include arguments "pool_layer", "separate_cls" in the description?

File: trax/models/research/funnel_transformer.py [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+      tl.Parallel(
+          None,
+          attention
+      ),                                # h', attention(...), mask
+      tl.Add(),                         # h'+attention(...), mask
+      tl.LayerNorm(),                   # funnel_activations, mask
+      tl.Parallel(
+          None,
+          tl.Fn('mask_max_pool',
+                _InternalMaxPool),
```
Comment (sebastianjaszczur) [inline trax/models/research/funnel_transformer.py:109]: This pool layer used for masks doesn't have pool_size/strides synchronized with pool layer for attention. So, setting pool_size/strides different than 2 will not work as I understand it.
Can you this pool layer to change mask according to pool_size?
(This comment applies also to _FunnelResidualBlock. )

File: trax/models/research/funnel_transformer.py [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+      ),                                # funnel_activations, mask'
+      feed_forward
```
Comment (sebastianjaszczur) [inline trax/models/research/funnel_transformer.py:111]: Why don't we have a residual connection here? The paper doesn't mention anything like that, and their PyTorch code seems to include it. If this is intended, can you write a brief comment about it?

File: trax/models/research/funnel_transformer.py [hunk 7]
```diff
@@ -0,0 +1,269 @@
...
+  )
+
+
+def FunnelTransformerEncoder(vocab_size,
```
Comment (sebastianjaszczur) [inline trax/models/research/funnel_transformer.py:115]: Can you add documentation to those models, along with argument description like in _FunnelBlock? This is applicable also to _FunnelResidualBlock and FunnelTransformer .

File: trax/models/research/funnel_transformer.py [hunk 8]
```diff
@@ -0,0 +1,269 @@
...
+  )
+
+
+def _FunnelResidualBlock(d_model, d_ff, n_heads,
+                         dropout, dropout_shared_axes, mode, ff_activation,
+                         pool_layer, pool_size, strides):
+  feed_forward = _FeedForwardBlock(
+      d_model, d_ff, dropout, dropout_shared_axes, mode, ff_activation)
+
+  dropout_ = tl.Dropout(
```
Comment (sebastianjaszczur) [inline trax/models/research/funnel_transformer.py:179]: Why do variables have "_" suffix here?

File: trax/models/research/funnel_transformer.py [hunk 9]
```diff
@@ -0,0 +1,269 @@
...
+      rate=dropout, shared_axes=dropout_shared_axes, mode=mode)
+
+  attn_ = tl.AttentionQKV(d_model, n_heads=n_heads, dropout=dropout, mode=mode)
+
+  pooling_ = PoolLayer(pool_layer, pool_size, strides)
+
+  return [
+      tl.Parallel(tl.Branch(pooling_, None), None),
```
Comment (sebastianjaszczur) [inline trax/models/research/funnel_transformer.py:187]: It's not very important, but can this be replaced by Select + pooling? I think it will be clearer.

File: trax/models/research/funnel_transformer.py [hunk 10]
```diff
@@ -0,0 +1,269 @@
...
+              'mask_max_pool', _InternalMaxPool)),
+          dropout_
+      ),
+      tl.Residual(
+          feed_forward
+      )
+  ]
+
+
+def FunnelTransformer(vocab_size,
```
Comment (sebastianjaszczur) [inline trax/models/research/funnel_transformer.py:202]: Can we rename it to FunnelTransformerDecoder, to keep naming consistent with models/transformer.py ? It seems closer to TransformerDecoder than Transformer, since the former outputs an embedding per token (like this Funnel class) and the latter predicts a class per token.

File: trax/models/research/funnel_transformer_test.py [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+from absl.testing import parameterized
+
+from trax import layers as tl, shapes
+from trax.models.research.funnel_transformer import PoolLayer, \
+  _FunnelResidualBlock, \
+  FunnelTransformerEncoder, \
+  FunnelTransformer
+
+
+class FunnelTransformerTest(parameterized.TestCase):
```
Comment (sebastianjaszczur) [inline trax/models/research/funnel_transformer_test.py:30]: Can we change all tests to use, approximately, the smallest possible models? E.g. d_model=8, d_ff=8, n_layers=2 etc. This would speed up tests very significantly while testing the same functionality.

**Review by sebastianjaszczur — CHANGES_REQUESTED** [review]
Thank you for the pull request, it looks good overall!
I have some important comments about the code, however.

Also, there are some linter errors; Afroz has shared the config we use for linter in https://github.com/google/trax/pull/1182 (this will be merged later). Can you run linter with this config and fix errors?

File: trax/models/research/funnel_transformer.py (hunk 1 above)
Comment (syzymon) [reply trax/models/research/funnel_transformer.py]: done

File: trax/models/research/funnel_transformer_test.py (hunk 11 above)
Comment (syzymon) [reply trax/models/research/funnel_transformer_test.py:28]: I have changed these parameters in tests, now they run much faster.

File: trax/models/research/funnel_transformer.py (hunk 3 above)
Comment (shadowatyyy) [reply trax/models/research/funnel_transformer.py:66]: done

File: trax/models/research/funnel_transformer.py (hunk 2 above)
Comment (shadowatyyy) [reply trax/models/research/funnel_transformer.py:64]: Asserting that input length is divisible by pool_size would be fine for FunnelTransformer with upsampling, but we use the same pooler for the version without it. We could modify pooler to consider whether its output will be upsampled later, but I'm not sure if it's worth it.

File: trax/models/research/funnel_transformer.py (hunk 10 above)
Comment (syzymon) [reply trax/models/research/funnel_transformer.py:215]: As per our previous discussion, we changed the `FunnelTransformer` to output token-level categorical distribution over vocab instead of embeddings, which makes it useful for example as a BERT.

File: trax/models/research/funnel_transformer.py (hunk 9 above)
Comment (syzymon) [reply trax/models/research/funnel_transformer.py:201]: This Parallel actually looks like a no-op so I removed it, but I would prefer not to apply pooling inside the residual with attention (Select is used there to split into Q, K, V).

File: trax/models/research/funnel_transformer.py (hunk 4 above)
Comment (syzymon) [reply trax/models/research/funnel_transformer.py:80]: done

File: trax/models/research/funnel_transformer.py (hunk 8 above)
Comment (syzymon) [reply trax/models/research/funnel_transformer.py]: Fixed - `dropout_` layer variable name was taken from the original TransformerEncoder to avoid shadowing `dropout` rate argument, replaced by a more meaningful `hidden_dropout`.

File: trax/models/research/funnel_transformer.py (hunk 7 above)
Comment (syzymon) [reply trax/models/research/funnel_transformer.py:135]: The docstrings have been added.

File: trax/models/research/funnel_transformer.py (hunk 6 above)
Comment (syzymon) [reply trax/models/research/funnel_transformer.py]: The paper authors are not clear on this indeed, but they mention:

> To inherit the high capacity and optimization advantages of the Transformer architecture, the proposed model keeps the same overall skeleton of interleaved S-Attn and P-FFN sub-modules wrapped by residual connection and layer normalization.

We decided to merge `_FunnelBlock` and `_FunnelResidualBlock` into one, with two residuals and two layer norms, similarly to the original PyTorch code (the only difference is they have a LN at the end of P-FFN, and Trax P-FFN seems to have one at the beginning, so we make up for it with a LN at the very beginning of the funnel block).

File: trax/models/research/funnel_transformer.py (hunk 2 above)
Comment (sebastianjaszczur) [reply trax/models/research/funnel_transformer.py:64]: What could be done fairly easily is to insert in _Upsample function check that "if long.shape[1] % short.shape[1] != 0: raise ValueError('message')." Then you don't need to modify anything but _Upsample function; downsampling isn't touched at all, and we can be sure that upsampling works correctly.
The current issue I have with this _Upsample function is that it computes wrong results (see my counterexample), and I think that throwing an exception is much better than silently returning wrong results.

Adding this check is also much easier than correcting the implementation - this would involve passing a pool_size/stride to _Upsample to be used in place of 'factor', and some padding (see the counterexample in my previous comment). I think that correcting the implementation may not be worth the effort, but adding an assert is worth it.

(Also, while it isn't necessary, I would consider adding this kind of assert even during downsampling. Let's consider the case when we have only downsampling, with pool_size=2, and input length not divisible by 2. Current downsampling simply throws away the last token (due to a padding "valid", which is the default), which is kind of a strange behaviour.)

**Review by sebastianjaszczur — CHANGES_REQUESTED** [review]
Thanks! I only wanted to follow up on implementation of _Upsample. Everything else looks good to me.
